### ISSUE TITLE
no-use-before-define false positive with named exports

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#9481]
**Tell us about your environment**

* **ESLint Version:** v4.8.0
* **Node Version:** v8.5.0
* **~~npm~~Yarn Version:** v1.0.2

**What parser (default, Babel-ESLint, etc.) are you using?** Default

(Eliding the rest...)

-----

This is a followup to [this comment in #7995](https://github.com/eslint/eslint/pull/7995#issuecomment-284568641) (and thus is not a dupe of #7858). I'd like to see an option added to `no-use-before-define` called `namedExports`, where:

- If `namedExports` is `true` (default), named exports are checked as normal.
- If `namedExports` is `false`, named exports are not checked.
- This has no effect with default exports.

Here's some examples:

```js
// Good
/* eslint no-use-before-define: ["error", {"namedExports": true}] */
const hasOwn = Object.prototype.hasOwnProperty
export {hasOwn}
```

```js
// Bad
/* eslint no-use-before-define: ["error", {"namedExports": true}] */
export {hasOwn}
const hasOwn = Object.prototype.hasOwnProperty
```

```js
// Good
/* eslint no-use-before-define: ["error", {"namedExports": false}] */
const hasOwn = Object.prototype.hasOwnProperty
export {hasOwn}
```

```js
// Good
/* eslint no-use-before-define: ["error", {"namedExports": false}] */
export {hasOwn}
const hasOwn = Object.prototype.hasOwnProperty
```

**Comment by Aladdin-ADD** [comment]
if I'm not misunderstanding, you are suggesting a new option to allow:
```js
// Good
/* eslint no-use-before-define: ["error", {"namedExports": false}] */
export {hasOwn}
const hasOwn = Object.prototype.hasOwnProperty
```

is there a special reason? (in my view, these code are mostly problematical.)

**Comment by isiahmeadows** [comment]
@Aladdin-ADD 

ES modules export references, not values. A reference is never directly evaluated, hence why `typeof foo === "undefined"` works even when `foo` is not yet defined in the scope. Additionally, references need not yet exist in the scope, but the reference may be defined later, whether programmatically or lexically. Here's such an example with `typeof`, which should never throw:

```js
// Note: this is a script.
;(global => {
    "use strict"
    function foo() {
        return typeof bar === "undefined"
    }

    let count = 0

    // Note: `bar` is not defined
    if (foo() === true) count++
    global.bar = 1

    if (foo() === false) count++

    if (count !== 2) throw "fail"
})(this)
```

Here's another example with `let`, that works very similarly, but causes the TDZ to be hit instead.

```js
// Note: this is a script.
;(global => {
    "use strict"
    function foo() {
        return typeof bar === "undefined"
    }

    let count = 0

    // Note: `bar` is defined, but not initialized
    if (foo() === true) count++
    let bar = 1

    if (foo() === false) count++

    if (count !== 2) throw "fail"
})(this)
```

If you were to export an undefined variable, it'd create a known reference error, but by spec, it will have to be delayed until the use site, since the reference is never evaluated until it's actually used. Here's an example of a fully statically analyzable case, where it should only throw if `doesNotExist` is a global, and it should never throw `"fail"`:

```js
// foo.js
if (typeof doesNotExist !== "undefined") throw "exists"
export {doesNotExist}

// bar.js
import {doesNotExist} from "./foo.js"
// It reaches here only if `doesNotExist` isn't a pre-defined global.

let hit = false

try {
    doesNotExist;
} catch (e) {
    hit = true
}

// This should never throw
if (!hit) throw "fail"
```

It's extraordinarily subtle, but that's why such exports work.

**Comment by isiahmeadows** [comment]
@Aladdin-ADD 

> is there a special reason?

See [this comment](https://github.com/eslint/eslint/pull/7995#issuecomment-276050651), which is basically my use case.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
